src/coreclr/jit/importer.cpp

```diff
@@ -13675,53 +13675,37 @@ void Compiler::impImportBlockCode(BasicBlock* block)
                 goto APPEND;
 
             case CEE_LDELEMA:
+            {
                 assertImp(sz == sizeof(unsigned));
 
                 _impResolveToken(CORINFO_TOKENKIND_Class);
 
                 JITDUMP(" %08X", resolvedToken.token);
 
                 ldelemClsHnd = resolvedToken.hClass;
+                lclTyp       = JITtype2varType(info.compCompHnd->asCorInfoType(ldelemClsHnd));
 
-                // If it's a value class array we just do a simple address-of
-                if (eeIsValueClass(ldelemClsHnd))
+                // If it's a value class / pointer array, or a readonly access, we don't need a type check.
+                if ((lclTyp != TYP_REF) || ((prefixFlags & PREFIX_READONLY) != 0))
```

Comment:
We can also apply `impCanSkipCovariantStoreCheck` to skip the helper, similar to how we do it in `CEE_STELEM_REF`.

For example, we can pass modify `impCanSkipCovariantStoreCheck` to treat `NULL` `value` as the ldelema case.